File: app/src/main/java/com/example/adam/myapplication/data/db/task/TaskRepository.java
package com.example.adam.myapplication.data.db.task;

import android.arch.lifecycle.LiveData;
import android.content.Context;
import android.os.AsyncTask;

import com.example.adam.myapplication.data.objects.Task;

import java.util.Date;
import java.util.List;
import java.util.concurrent.ExecutionException;

public class TaskRepository {

    private static TaskDao dao;

    public TaskRepository(Context context) {
        dao = TaskDatabase.getInstance(context).getTaskDao();
    }

    public LiveData<List<Task>> getAll() {
        return dao.getAll();
    }

    public List<Task> getAllFromDate(Date start, Date end) {
        GetTask task = new GetTask(start, end);

        try {
            return task.execute().get();

        } catch (ExecutionException e) {
            e.printStackTrace();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        return null;
    }

    public void insert(final Task task) {
        InsertThread insertThread = new InsertThread(task, dao);
        insertThread.start();
    }

    public void insert(List<Task> tasks) {
        InsertManyThread insertManyThread = new InsertManyThread(tasks, dao);
        insertManyThread.start();
    }

    public void update(Task task){
        UpdateThread updateThread = new UpdateThread(task, dao);
        updateThread.start();
    }

    public void delete(Task task) {
        DeleteThread deleteThread = new DeleteThread(task, dao);
        deleteThread.start();
    }

    private static class GetTask extends AsyncTask<Void, Void, List<Task>> {

        private Date start;
        private Date end;

        GetTask(Date start, Date end) {
            this.start = start;
            this.end = end;
        }

        @Override
        protected List<Task> doInBackground(Void... voids) {
            return dao.getAllFromTimeWindow(start, end);
        }
    }

    private class InsertThread extends Thread {
        private Task task;
        private TaskDao dao;

        InsertThread(Task task, TaskDao dao) {
            this.task = task;
            this.dao = dao;
        }

        @Override
        public void run() {
            if (task != null)
                dao.insert(task);
        }
    }

    private class InsertManyThread extends Thread {
        private List<Task> tasks;
        private TaskDao dao;

        InsertManyThread(List<Task> tasks, TaskDao dao) {
            this.tasks = tasks;
            this.dao = dao;
        }

        @Override
        public void run() {
            for (Task task : tasks)
                dao.insert(task);
        }
    }

    private class UpdateThread extends Thread {
        private Task task;
        private TaskDao dao;

        UpdateThread(Task task, TaskDao dao){
            this.task = task;
            this.dao = dao;
        }

        @Override
        public void run() {
            dao.update(task);
        }
    }

    private class DeleteThread extends Thread {
        private Task task;
        private TaskDao dao;

        DeleteThread(Task task, TaskDao dao) {
            this.task = task;
            this.dao = dao;
        }

        @Override
        public void run() {
            dao.delete(task);
        }
    }
}

File: app/src/main/java/com/example/adam/myapplication/data/db/task/TaskRepository.kt
package com.example.adam.myapplication.data.db.task

import android.arch.lifecycle.LiveData
import android.content.Context
import android.os.AsyncTask

import com.example.adam.myapplication.data.objects.Task

import java.util.Date
import java.util.concurrent.ExecutionException

class TaskRepository(context: Context) {

    companion object {
        private lateinit var dao: TaskDao
    }

    init {
        dao = TaskDatabase.getInstance(context).taskDao
    }

    fun getAll(): LiveData<List<Task>> {
        return dao.getAll()
    }

    fun getAllFromDate(start: Date, end: Date): List<Task>? {
        val task = GetTask(start, end)

        try {
            return task.execute().get()

        } catch (e: ExecutionException) {
            e.printStackTrace()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

        return null
    }

    fun insert(task: Task) {
        val insertThread = InsertThread(task, dao)
        insertThread.start()
    }

    fun insert(tasks: List<Task>) {
        val insertManyThread = InsertManyThread(tasks, dao)
        insertManyThread.start()
    }

    fun update(task: Task) {
        val updateThread = UpdateThread(task, dao)
        updateThread.start()
    }

    fun delete(task: Task) {
        val deleteThread = DeleteThread(task, dao)
        deleteThread.start()
    }

    private class GetTask internal constructor(private val start: Date, private val end: Date) : AsyncTask<Void, Void, List<Task>>() {

        override fun doInBackground(vararg voids: Void): List<Task> {
            return dao.getAllFromTimeWindow(start, end)
        }
    }

    private inner class InsertThread internal constructor(private val task: Task?, private val dao: TaskDao) : Thread() {

        override fun run() {
            if (task != null)
                dao.insert(task)
        }
    }

    private inner class InsertManyThread internal constructor(private val tasks: List<Task>, private val dao: TaskDao) : Thread() {

        override fun run() {
            for (task in tasks)
                dao.insert(task)
        }
    }

    private inner class UpdateThread internal constructor(private val task: Task, private val dao: TaskDao) : Thread() {

        override fun run() {
            dao.update(task)
        }
    }

    private inner class DeleteThread internal constructor(private val task: Task, private val dao: TaskDao) : Thread() {

        override fun run() {
            dao.delete(task)
        }
    }
}